pkg/diagnose/runner.go

```diff
@@ -339,41 +339,81 @@ func requestDiagnosesFromAgentProcess(diagCfg diagnosis.Config) ([]diagnosis.Dia
 	return diagnoses, nil
 }
 
+// RunInCLIProcess run diagnoses in the CLI process.
+func RunInCLIProcess(diagCfg diagnosis.Config, deps SuitesDepsInCLIProcess) ([]diagnosis.Diagnoses, error) {
+	return run(diagCfg, func() []diagnosis.Suite {
+		return buildSuites(diagCfg, func() []diagnosis.Diagnosis {
+			return diagnoseChecksInCLIProcess(diagCfg, deps.senderManager, deps.secretResolver, deps.wmeta, deps.AC)
+		})
+	})
+}
+
 // Run runs diagnoses.
-func Run(diagCfg diagnosis.Config, deps SuitesDeps) ([]diagnosis.Diagnoses, error) {
+func run(diagCfg diagnosis.Config, getSuites func() []diagnosis.Suite) ([]diagnosis.Diagnoses, error) {
 	// Make remote call to get diagnoses
 	if !diagCfg.RunLocal {
 		return requestDiagnosesFromAgentProcess(diagCfg)
 	}
 
 	// Collect local diagnoses
-	diagnoses, err := getDiagnosesFromCurrentProcess(diagCfg, getSuites(diagCfg, deps))
+	diagnoses, err := getDiagnosesFromCurrentProcess(diagCfg, getSuites())
 	if err != nil {
 		return nil, err
 	}
 
 	return diagnoses, nil
 }
 
-// Enumerate registered Diagnose suites and get their diagnoses
+// RunInAgentProcess runs diagnoses in the Agent process.
+func RunInAgentProcess(diagCfg diagnosis.Config, deps SuitesDepsInAgentProcess) ([]diagnosis.Diagnoses, error) {
+	// Collect local diagnoses
+	suites := buildSuites(diagCfg, func() []diagnosis.Diagnosis {
+		return diagnoseChecksInAgentProcess(deps.collector)
+	})
+
+	return getDiagnosesFromCurrentProcess(diagCfg, suites)
+}
+
+// RunStdOutInAgentProcess enumerates registered Diagnose suites and get their diagnoses
 // for human consumption
-//
-//nolint:revive // TODO(CINT) Fix revive linter
-func RunStdOut(w io.Writer, diagCfg diagnosis.Config, deps SuitesDeps) error {
+func RunStdOutInAgentProcess(w io.Writer, diagCfg diagnosis.Config, deps SuitesDepsInAgentProcess) error {
+	return runStdOut(w, diagCfg, func(diagCfg diagnosis.Config) ([]diagnosis.Diagnoses, error) {
+		return RunInAgentProcess(diagCfg, deps)
+	})
+}
+
+// RunStdOutInCLIProcess enumerates registered Diagnose suites and get their diagnoses
+// for human consumption
+func RunStdOutInCLIProcess(w io.Writer, diagCfg diagnosis.Config, deps SuitesDepsInCLIProcess) error {
+	return runStdOut(w, diagCfg, func(diagCfg diagnosis.Config) ([]diagnosis.Diagnoses, error) {
+		return RunInCLIProcess(diagCfg, deps)
+	})
+}
+
+// RunStdOutAutodiscovery run the check "connectivity-datadog-autodiscovery"
+func RunStdOutAutodiscovery(w io.Writer, diagCfg diagnosis.Config) error {
```

Comment:
Why we need an explicit function. Using this approach would require all suites to have a dedicated function. Previous include filter was sufficient.

Reply:
Previously the code uses `RunStdOut`. Calling `RunStdOut(w io.Writer, diagCfg diagnosis.Config, deps SuitesDeps)` requires or not `deps` depending on the check name. The user has to look at the implementation to know if a check needs `deps` or not (To be more precise if they can use optional.None or not for fields inside `SuitesDeps`). From an API perspective, I think hiding the implementation details is important.

When you designed this code, the choice you made make a lot of senses. it works well because globals hide the fact that different checks and code path require different dependencies. 

If you are worry about having one function per check, I can rename to something like `RunStdOutExceptDatadogCheck` (Naming should be improve).  

Another solution would be to have `RunStdOutCheck` taking the list of checks to run.

What are you thoughts?

Reply:
Let's brainstorm it on zoom